packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -838,7 +845,17 @@ export abstract class BucketBase extends Resource implements IBucket {
    * https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html
    */
   public addEventNotification(event: EventType, dest: IBucketNotificationDestination, ...filters: NotificationKeyFilter[]) {
-    this.notifications.addNotification(event, dest, ...filters);
+    this.withNotifications(notifications => notifications.addNotification(event, dest, ...filters));
+  }
+
+  private withNotifications(cb: (notifications: BucketNotifications) => void) {
```

Comment:
Im not sure I see why this `with` pattern is needed. 
The `BucketNotifications` class already lazily creates a resource only if notifications are added. 

You can just pass `notificationsHandlerRole` as a property to `BucketBase`, and keep the initialization of `BucketNotifications` in the base constructor.

Reply:
The problem is: `BucketBase` gets a `ResourceProps` as a constructor argument. If I replace it with a new `BucketBaseProps` interface, that has the new property, jsii complains with:

>err  - INITIALIZER @aws-cdk/aws-s3.BucketBase.<initializer>: argument props, takes Optional<@aws-cdk/aws-s3.BucketBaseProps> (formerly Optional<@aws-cdk/core.ResourceProps>): @aws-cdk/core.ResourceProps does not extend @aws-cdk/aws-s3.BucketBaseProps [incompatible-argument:@aws-cdk/aws-s3.BucketBase.<initializer>]

So this is how I'm working around the issue. But I'm not very happy with it, either. If you have a better idea, I'd be happy to consider it.

Reply:
Hmm. Yeah I guess this breaks compatibility for java and such. Ok 👍